File: lab_10/calculator.cpp
#include <iostream>
using namespace std;

class Calculator {
public:
    int add(int a, int b) {
        return a + b;
    }
};

class SubtractionCalculator : public Calculator {
public:
    int subtract(int a, int b) {
        return a - b;
    }
};

class MultiplicationCalculator : public SubtractionCalculator {
public:
    int multiply(int a, int b) {
        return a * b;
    }
};

class DivisionCalculator : public MultiplicationCalculator {
public:
    int divide(int a, int b) {
        if(b == 0) {
            cout << "Error: Cannot divide by zero." << endl;
            return 0;
        }
        return a / b;
    }
};

int main() {
    DivisionCalculator mulCalc;

    cout << "Addition (10 + 5): " << mulCalc.add(10, 5) << endl;

    cout << "Subtraction (10 - 5): " << mulCalc.subtract(10, 5) << endl;

    cout << "Multiplication (10 * 5): " << mulCalc.multiply(10, 5) << endl;

    cout << "Division (10 / 5): " << mulCalc.divide(10, 5) << endl;

    cout << "Division (10 / 0): " << mulCalc.divide(10, 0) << endl;

    return 0;
}

File: lab_10/calculator.py
class Calculator:
    def add(self, a, b):
        return a + b

class SubtractionCalculator(Calculator):
    def subtract(self, a, b):
        return a - b

class MultiplicationCalculator(SubtractionCalculator):
    def multiply(self, a, b):
        return a * b

class DivisionCalculator(MultiplicationCalculator):
    def divide(self, a, b):
        if b == 0:
            return "Cannot divide by zero"
        return a / b

if __name__ == "__main__":
    mul_calc = DivisionCalculator()

    print(f"Addition (10 + 5): {mul_calc.add(10, 5)}")

    print(f"Subtraction (10 - 5): {mul_calc.subtract(10, 5)}")

    print(f"Multiplication (10 * 5): {mul_calc.multiply(10, 5)}")

    print(f"Division (10 / 5): {mul_calc.divide(10, 5)}")

    print(f"Division (10 / 0): {mul_calc.divide(10, 0)}")

